### DISCUSSION TITLE
decreasing the amount of zoom in zoomToElements()

### DISCUSSION THREAD

**Discussion by kardoka** [iTwin/itwinjs-core#4492] [Q&A] [answered]
Greetings,

I am experimenting with [`zoomToElements`](https://www.itwinjs.org/reference/core-frontend/views/viewport/zoomtoelements/) and I would want to make it zoom in less. I have set the [`MarginPercent`](https://www.itwinjs.org/reference/core-frontend/views/marginpercent/) to 0.25 for all sides (the maximum allowed value), which helps, but I would like to zoom out more.

I have tried out various parameters for the `zoomToElements` method, including passing a `Vector3` to [`GlobalAlignmentOptions`](https://www.itwinjs.org/reference/core-frontend/views/globalalignmentoptions/), but that does not seem to help.

What are my options here? Could it be possible to achieve less zoom with `zoomToElements` or do I have to use some other approach?

**Comment by pmconne** [comment, marked as answer]
> I would like to zoom out more

You simply mean, you want `MarginPercent` to permit values > 0.25?
Or, do you want some other customization? If so, what?

**Reply by kardoka** [reply]
> You simply mean, you want `MarginPercent` to permit values > 0.25?

If that would be possible, it would achieve what I am looking for.

But if that is not possible I would be looking forward to other methods of accomplishing the same thing.

**Reply by pmconne** [reply]
I assume you have (and want) the camera turned off, as the margin is not applicable to perspective views, and that you want the same margins to apply to any element regardless of the element's size. Please correct me if I misunderstand.

@kabentley @bbastings do you recall why `lookAtVolume` prohibits a margin > 0.25 on each side?

**Reply by kabentley** [reply]
no, I don't remember. Seems like it could be much larger without problems. What it means is: take the width/height of the calculated volume and expand each side (left, right, top, bottom) by that distance times its margin percent. If you use 1.0 for each value, elements would take up 1/3 of the view.

**Reply by pmconne** [reply]
Thanks. @kardoka I will submit a PR. In the meantime you ought to be able to assign values greater than .25 to the properties of `MarginPercent` after creating it -- the clamping is only applied in the constructor.

**Reply by kabentley** [reply]
in the meantime, you can just pass:

```ts
    zoomToElements(ids, { marginPercent: { left: .5, right: .5, top: .5, bottom: .5 } });
```
and skip the test in the `MarginPercent` class.

(or use Paul's suggestion of creating a MarginPercent and then modifying it.)

**Reply by pmconne** [reply]
> `zoomToElements(ids, { marginPercent: { left: .5, right: .5, top: .5, bottom: .5 } });`

That won't work; the sums left+right and top+bottom must be less than 1.

**Reply by williamkbentley** [reply]
[Example](https://www.itwinjs.org/sandboxes/williamkbentley/zoom-to-margin-percent). Adjust the slider to set `marginPercent` then click the button to do the zoom.

**Reply by kabentley** [reply]
@williamkbentley very cool, thanks!

**Reply by kardoka** [reply]
Thank you all for the suggestions, I will try your solutions out.

**Reply by kardoka** [reply]
> [Example](https://www.itwinjs.org/sandboxes/williamkbentley/zoom-to-margin-percent). Adjust the slider to set `marginPercent` then click the button to do the zoom.

This solution works! Thank you very much.

**Reply by pmconne** [reply]
@kardoka please do not rely on the current behavior when the margin is greater than 0.25. The math comes out wrong.
Example: margin=0.25
![image](https://user-images.githubusercontent.com/22944042/195899493-e41248ef-c89b-4111-8f33-7784d1ea6a21.png)

margin=0.49
![image](https://user-images.githubusercontent.com/22944042/195899573-4c1959ab-36b0-4e36-b96a-8c75ea9953a2.png)

**Comment by pmconne** [comment]
#4496 adds a more reliable way to apply margins when zooming to elements.
